[11.0] [MIG] base_rest, base_rest_demo: Migration to 11.0

lmignon opened the pull request.

lmignon committed:
[ADD] New generic module du define REST services

lmignon committed:
[IMP] base_rest: Initialize session lang from the lang found into the HTTP headers if provided

lmignon committed:
[FIX] base_rest: Remove referene to api-key

RobinetDenisAcsone committed:
[FIX] base_rest: Correct import of datetime

RobinetDenisAcsone committed:
[FIX] base_rest: Remove the call to _validate_list since the issue (https://github.com/pyeve/cerberus/issues/325) is resolved

lmignon committed:
[IMP] base_rest: Simplify controller dev and services registration

lmignon committed:
[IMP] base_rest: allow to specify auth, csrf and cors to use for the default routes

lmignon committed:
[FIX] base_rest: fix README.rst

lmignon committed:
[FIX] base_rest: BaseRestService must be an AbstractComponent

lmignon committed:
[FIX] base_rest: remove dead code

lmignon committed:
[FIX] base_rest: typo in log message

lmignon committed:
[FIX] base_rest: properly decode request data provided as json

sebastienbeau committed:
[REF] refactor code in order to use generic code from base_rest instead of custom one from shopinvader

sebastienbeau committed:
[FIX] fix comment done by Laurent, move generic common test in base_rest

sebastienbeau committed:
[IMP] python do not parse correctly data structured query string, add the dependency on pyquerystring that do it correctly

lmignon committed:
[FIX]base_rest: properly declare external dependencies

lmignon committed:
[IMP] base_rest: Correctly process the http header 'Accept-language' to initialize the language into the Odoo context

lmignon committed:
[IMP] base_rest: Always log the stack trace in case of exception

bealdav committed:
Update README.rst

lmignon committed:
Move from https://github.com/akretion/odoo-shopinvader

lmignon committed:
Add swagger frontend and generate OpenAPI definitions of services

lmignon committed:
pep8

lmignon committed:
[FIX] base_rest: Remove the word 'invader' from the code

lmignon committed:
[FIX] base_rest: Typo

lmignon committed:
[FIX] base_rest: Don't parse integer if already an integer

lmignon committed:
[IMP] base_rest: Add services description
Allow to specify a description into the services that will be used into the openapi documentation.
It's not possible to use the docstring put on the service component as description since this information is lost when the final class is generated once the registery is loaded

lmignon committed:
[FIX] base_rest: allow the use of multi part path to register services

lmignon committed:
[FIX] base_rest: Process SessionExpiredException as real exception
By default in case of SessionExpiredException, Odoo return the login page with HTTP Status 200. For REST services we must return the real exception IOW 401 Unauthorize

lmignon committed:
[FIX] base_rest: Id is by convention an integer:
Into the openapi documentation, declare 'id' as an integer path parameter

lmignon committed:
[FIX] base_rest: Fix openapi documentation for HTTP GET methods:
In HTTP GET methods, the paramters are not provided as a json document. The parameters are http query parameters. Therefore we need to translate the jsonschema defined for services available via HTTP GET into an openapi definition of the query paramters.

lmignon committed:
gitignore

lmignon committed:
[FIX] base_rest: lint

lmignon committed:
[IMP] base_rest_demo: New module for tests and demo

lmignon committed:
[FIX] base_rest: Fix RestController Factory
To avoid to redefine the generic routes in all our RestController, the routes was defined in the RestController base classe. Unfortunately even if the path into the routes was modified into the class factrory the definition of the generic routs was the same for all the Controllers inheriting from RestController. To avoid to share the definitions of these generic methods, we override the class factory into the metaclass and generate a new definition of these methods for each classes inheriting from RestController.
In the same time we must also remove the RestConroller from the controller since it's an abstract controller.

lmignon committed:
[IMP] base_rest: Improve methods naming

lmignon committed:
[IMP] base_rest, base_rest_demo: Add tests

lmignon committed:
11.0
[skip ci]

lmignon committed:
[MIG] base_rest, base_rest_demo: Migration to 11.0

lmignon committed:
[FIX] base_rest: preserve list order into the openapi def

lmignon committed:
[FIX] base_rest: Add missing methods into openapi def for http methods put and del

lmignon committed:
[MIG] base_rest: In python3 inspect
In Py3 a function becomes a method only once the class def is instanciated.

lmignon committed:
[FIX] config paramters are only available in sudo

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[FIX] base_rest: preserve list order into the openapi def

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[FIX] base_rest: in PY3 request.query_string is bytes and pyquerystring parser expect test

lmignon committed:
[IMP] base_rest: Add menu and icons to access to the swagger interface

lmignon committed:
[IMP] base_rest: open swagger UI with the specs specified into the url
When navigating from one spec to an other, the page url is updated to include the name of the displayed page as paramter. This change takes into account this parameter when generating the swagger page to reopen the page with the selected spec open by default. In the same time, this change avoid a stacktrace when the api-docs page is called with paramters

lmignon committed:
[IMP] base_rest, base_rest_demo: By default get method is called without parameter

lmignon committed:
Merge branch '10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[IMP] base_rest: Sort method by method and path into the swagger UI

lmignon committed:
Merge branch '10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[IMP] base_rest: Include stacktrace into repsonse in case of error if log level is DEBUG

lmignon committed:
Merge branch '10.0-base_rest_swagger' into 11.0-from-10.0

sbidoul commented on base_rest/http.py:
```diff
@@ -43,11 +43,18 @@ def wrapJsonException(exception):
             'name': escape(exception.name),
             'description': exception.get_description(environ)
+            }
+        if _logger.isEnabledFor(logging.DEBUG):
```
This sounds dangerous. Enabling `--log-level=debug` globally will have the (surprising IMO) side effect of enabling stack traces in base_rest (which is often considered a security issue).

I'd prefer an option in the config file, that we could then manage with `server_environment`.

lmignon replied:
@sbidoul Thank you for this insightful comment. I'll change the code to use an option into the config file.

lmignon replied:
@sbidoul What do you think about https://github.com/OCA/rest-framework/pull/5/commits/317dcb4a91ae049af6c6778c6e6ef54222f31f77

sbidoul replied:
I made a couple of comments on the commit.

lmignon committed:
[IMP] base_rest: Returns stack_trace only if base_rest is in dev_mode
Using the log level to enable the stack trace into the response is too dangerous. We must avoid to mix log level and development mode

lmignon committed:
[IMP] base_rest: Put specific config option into a dedicated section '[base_rest]'

lmignon committed:
[FIX] base_rest: Validation errors must be part of the UserError

lmignon committed:
[FIX] base_rest: output validation error is a system error
not a user error. Because it's the developer's fault, not the user's

lmignon committed:
[FIX] base_rest_demo: Typo

lmignon committed:
[IMP] base_rest: include error description into the json response in case of UserError or ValidationError

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[IMP] base_rest_demo: Add tests for exceptions

lmignon committed:
[FIX] base_rest: decorators must return the decorated method

lmignon committed:
[IMP] base_rest, base_rest_demo: Add support for 'enum' in request/response schema definition

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[FIX] base_rest, base_rest_demo: Fix exceptions handling

lmignon committed:
[FIX] base_rest, base_rest_demo: Remove leading whitespace from every lines into openapi service descriptions and method summaries

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

lmignon committed:
[IMP] base_rest: Add button to return to the web client
Refactor UI to follow Odoo convention

lmignon committed:
[IMP] base_rest: Add button to return to the web client
Refactor UI to follow Odoo convention

lmignon committed:
Typo

lmignon committed:
Merge remote-tracking branch 'origin/10.0-base_rest_swagger' into 11.0-from-10.0

sbidoul commented on .travis.yml:
```diff
@@ -3,7 +3,7 @@ sudo: false
 
 python:
-  - "2.7.13"
+  - "3.6"
```
For 11.0 we need to test on 3.5 which is the minimum version supported by Odoo.

sbidoul commented on setup/base_rest_demo/odoo/addons/__init__.py:
```diff
@@ -0,0 +1 @@
+__import__('pkg_resources').declare_namespace(__name__)
```
no init.py for 11

sbidoul reviewed (approved):
A couple of minor changes, otherwise +1

lmignon committed:
[FIX] setup.py for PY3 (no __init__.py)

lmignon committed:
[FIX] travis.yml: For 11.0 we need to test on 3.5 which is the minimum version supported by Odoo.

lmignon commented:
Thank you @sbidoul  Changes done

lmignon committed:
Merge branch '11.0' into 11.0-from-10.0

lmignon committed:
[FIX] base_rest_demo: PY 3.5 compatibility
In python 3.5 json.loads only support to deserialize str instance, not
bytes. (As of 3.6 str, bytes or bytearray instane are supported)

lmignon committed:
[IMP] base_rest: Always decode request data from request charset

sebastienbeau commented:
Ok for me :+1:
